src/lib/react/useRoomConnection.ts

```diff
@@ -354,94 +369,148 @@ export default function useRoomConnection(
+                        dispatch({ type: "ROOM_CONNECTION_STATUS_CHANGED", payload: { roomConnectionStatus } });
+                    },
+                },
+                {
+                    eventName: "room_joined",
+                    listener: (e) => {
+                        const { localParticipant, remoteParticipants, waitingParticipants } =
+                            e.detail as RoomJoinedEvent;
+                        dispatch({
+                            type: "ROOM_JOINED",
+                            payload: { localParticipant, remoteParticipants, waitingParticipants },
+                        });
+                    },
+                },
+                {
+                    eventName: "screenshare_started",
+                    listener: (e) => {
+                        const { participantId, id, hasAudioTrack, stream } = e.detail as ScreenshareStartedEvent;
+                        dispatch({
+                            type: "SCREENSHARE_STARTED",
+                            payload: { participantId, id, hasAudioTrack, stream },
+                        });
+                    },
+                },
+                {
+                    eventName: "screenshare_stopped",
+                    listener: (e) => {
+                        const { participantId, id } = e.detail as ScreenshareStoppedEvent;
+                        dispatch({ type: "SCREENSHARE_STOPPED", payload: { participantId, id } });
+                    },
+                },
+                {
+                    eventName: "streaming_started",
+                    listener: (e) => {
+                        const { status, startedAt } = e.detail as StreamingState;
+                        dispatch({ type: "STREAMING_STARTED", payload: { status, startedAt } });
+                    },
+                },
+                {
+                    eventName: "streaming_stopped",
+                    listener: () => {
+                        dispatch({ type: "STREAMING_STOPPED" });
+                    },
+                },
+                {
+                    eventName: "waiting_participant_joined",
+                    listener: (e) => {
+                        const { participantId, displayName } = e.detail as WaitingParticipantJoinedEvent;
+                        dispatch({ type: "WAITING_PARTICIPANT_JOINED", payload: { participantId, displayName } });
+                    },
+                },
+                {
+                    eventName: "waiting_participant_left",
+                    listener: (e) => {
+                        const { participantId } = e.detail as WaitingParticipantLeftEvent;
+                        dispatch({ type: "WAITING_PARTICIPANT_LEFT", payload: { participantId } });
+                    },
+                },
+            ] as { eventName: keyof RoomEventsMap; listener: RoomEventListener<keyof RoomEventsMap> }[],
+        []
```

Comment:
Same here, we should declare this type on the function, and not assert with `as`.

Reply:
This would also be nice, but I have troubles making TS happy with whatever typedef I try. It fails to match the event name + listener cb args. 😩 

Tried:

```ts
type EventListenerObject<K extends keyof RoomEventsMap> = {
    eventName: K;
    listener: (e: RoomEventsMap[K]) => void;
};

const eventListeners = React.useMemo(
    (): Array<EventListenerObject<keyof RoomEventsMap>> => [
        {
            eventName: "chat_message",
            listener: (e: CustomEvent<ChatMessage>) => {
                const chatMessage = e.detail;
                dispatch({ type: "CHAT_MESSAGE", payload: chatMessage });
            },
        },
        {
            eventName: "cloud_recording_started",
            listener: (e: CustomEvent<CloudRecordingState>) => {
                const { status, startedAt } = e.detail;
                dispatch({ type: "CLOUD_RECORDING_STARTED", payload: { status, startedAt } });
            },
        },
        // ...
    ],
    []
)
```
Complains on the listener:
```
      Type 'CustomEvent<CloudRecordingState>' is not assignable to type 'CustomEvent<ChatMessage>'.
        Type 'CloudRecordingState' is missing the following properties from type 'ChatMessage': senderId, timestamp, text (tsserver 2322)
```

```ts
// this was the initial one moved up from the type assert
type RoomEventListener<K extends keyof RoomEventsMap> = (e: RoomEventsMap[K]) => void;
type EventListenerObject = { eventName: keyof RoomEventsMap; listener: RoomEventListener<keyof RoomEventsMap> };
const eventListeners = React.useMemo(
    (): Array<EventListenerObject<keyof RoomEventsMap>> => [
        // ... same objects as above
    ],
    []
)
```

The TS error is the same here as above 😕 

Do you have any hint?

Reply:
I pushed my not-yet-working prototype in a new commit. You can either continue working on that or drop it.

Thanks for picking this up! 🙏

Reply:
Fixed the typings now. It infers correctly now.